src/plugins/intel_cpu/src/mkldnn_weights_cache.cpp

```diff
@@ -36,38 +36,38 @@ MKLDNNWeightsSharing::MKLDNNSharedMemory::Ptr MKLDNNWeightsSharing::findOrCreate
                             const std::string& key,
                             std::function<MKLDNNMemoryPtr(void)> create,
                             bool valid) {
-    std::unique_lock<std::mutex> lock(guard);
-    auto found = sharedWeights.find(key);
-
+    std::unique_lock<std::mutex> sharedMemoryLock;
     MKLDNNMemoryInfo::Ptr ptr;
     MKLDNNMemoryPtr newPtr;
-
-    if (found == sharedWeights.end()
-        || !((ptr = found->second) && (newPtr = ptr->sharedMemory.lock()))) {
-        newPtr = create();
-        ptr = std::make_shared<MKLDNNMemoryInfo>(newPtr, valid);
-        sharedWeights[key] = ptr;
-    }
-
-    return std::make_shared<MKLDNNSharedMemory>(ptr->valid.load(std::memory_order_relaxed)
-                                                ? std::unique_lock<std::mutex>(ptr->guard, std::defer_lock)
-                                                : std::unique_lock<std::mutex>(ptr->guard), ptr, newPtr);
+    do {
+        std::unique_lock<std::mutex> lock(guard);
+        auto found = sharedWeights.find(key);
+
+        if (found == sharedWeights.end()
+            || !((ptr = found->second) && (newPtr = ptr->sharedMemory.lock()))) {
+            newPtr = create();
+            ptr = std::make_shared<MKLDNNMemoryInfo>(newPtr, valid);
+            sharedWeights[key] = ptr;
+        }
```

Comment:
It seems this part we can extract to separate function and call it once outside the loop.

Reply:
Agree. As we discussed, the cycle is not really required.